File: Olivia/1119/Linked List Cycle II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head==null ||head.next==null){
            return null;
        }
        ListNode fast=head;
        ListNode slow=head;
        while(fast.next!=null &&fast.next.next!=null){
            slow=slow.next;
            fast=fast.next.next;
            if(slow==fast){
                slow=head;
                while(slow!=fast){
                    slow=slow.next;
                    fast=fast.next;
                }
                return slow;
            }
        }
        return null;
    }
}
//t=x+ny+k, 2t=x+my+k, so 2x+2n+2k=x+my+k, so x+k=(m-2n)y

File: Sen/Linked_List_Cycle_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head): # extra space: O(n)
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return None
        dic = {}
        while head:
            if head in dic:
                return head
            dic[head] = 0
            head = head.next
        return None

    def detectCycle(self, head): 
        if not head or not head.next:
            return None
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow is fast:
                slow = head
                while slow != fast:
                    slow = slow.next
                    fast = fast.next
                return slow
        return None
